pkg/client/bigquery.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+// BigQuery provides a client and dataset identifiers.
+type BigQuery struct {
+	projectID string
+	datasetID string
+	client    *bigquery.Client
+}
+
+// CheckpointEntry is the shape of an entry to the checkpoint table.
+type CheckpointEntry struct {
+	deliveryID string
+	createdAt  string
+}
+
+// FailureEventEntry is the shape of an entry to the failure_events table.
+type FailureEventEntry struct {
+	deliveryID string
+	createdAt  string
+}
+
+// NewBigQueryClient creates a new instance of a BigQuery client.
+func NewBigQueryClient(ctx context.Context, projectID, datasetID string, opts ...option.ClientOption) (*BigQuery, error) {
+	client, err := bigquery.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bigquery client: %w", err)
+	}
+
+	return &BigQuery{
+		projectID: projectID,
+		datasetID: datasetID,
+		client:    client,
+	}, nil
+}
+
+func (bq *BigQuery) Close() {
+	bq.client.Close()
+}
+
+// Check if an entry with a given delivery_id already exists in the events table, this attempts to prevent duplicate processing of events.
+// This is used by the webhook service.
+func (bq *BigQuery) DeliveryEventExists(ctx context.Context, eventsTableID, deliveryID string) (bool, error) {
+	res, err := bq.makeCountQuery(ctx, fmt.Sprintf("SELECT COUNT(1) FROM `%s.%s.%s` WHERE delivery_id = '%s'",
```

Comment:
You could use %q here instead of %s

```suggestion
	res, err := bq.makeCountQuery(ctx, fmt.Sprintf("SELECT COUNT(1) FROM `%s.%s.%s` WHERE delivery_id = %q",
```